### PR TITLE
Fffr

### PR DESCRIPTION
**Pull request by Sibras**

### CODE REVIEW

**Comment by anibali** [comment]
With https://github.com/anibali/tvl/pull/13/commits/56b5cd75163e5d6d4d8f30667297552d73c1a025 I get `Stream creation failed.` when using the primary GPU.

EDIT: I've found and fixed the problem, see https://github.com/anibali/tvl/pull/13#discussion_r299250024

File: tvl_backends/tvl-backends-fffr/ext/bindings/TvFFFrameReader.cpp [hunk 1]
```diff
@@ -169,17 +169,17 @@ int64_t TvFFFrameReader::read_frame_sequence(int64_t* offsets, int n_frames, uin
     // to get all of them at once.
     std::vector<int64_t> frameSequence(offsets, offsets + n_frames);
     const auto frames_vector = _stream->getNextFrameSequence(frameSequence);
-    int n_frames_read = frames_vector.size();
```
Comment (anibali) [inline tvl_backends/tvl-backends-fffr/ext/bindings/TvFFFrameReader.cpp]: What error handling scenario is being fixed by this change?

**Comment by anibali** [comment]
What error detection scenario is being fixed in https://github.com/anibali/tvl/pull/13/commits/7dd6c39b29479808c0d951fc5b9b2eb9c0ae6bba?

File: tvl_backends/tvl-backends-fffr/ext/bindings/TvFFFrameReader.cpp [hunk 2]
```diff
@@ -66,7 +58,7 @@ TvFFFrameReader::TvFFFrameReader(ImageAllocator* image_allocator, const std::str
         options.m_type = Ffr::DecodeType::Cuda;
         options.m_outputHost = false;
         options.m_format = Ffr::PixelFormat::Auto; // Keep pixel format the same
-        options.m_context = _context.get();
+        options.m_context = _contexts[gpu_index];
```
Comment (anibali) [inline tvl_backends/tvl-backends-fffr/ext/bindings/TvFFFrameReader.cpp:61]: Found the problem: this should be `_contexts[gpu_index].get()`

**Comment by anibali** [comment]
I've introduced just the multiple GPU fix in https://github.com/anibali/tvl/commit/4e55ff045ad1ee30b081628ae0c17ce9084450cb.

I haven't incorporated the change to read_frame_sequence yet as I would like to know the rationale behind it first. In fact, that should become its own separate issue/PR.

**Comment by Sibras** [comment]
The change for read_frame_sequence is because `if (frame == nullptr)` will never be true as the function doesnt return nullptr on failure it instead just returns fewer elements in the returned vector

**Comment by anibali** [comment]
I see. I'll add your change to that code tomorrow then.

**Pull request closed without merging**